server/server/src/securityStorage.ts

```diff
@@ -0,0 +1,89 @@
+//
+// Copyright © 2021 Anticrm Platform Contributors.
+//
+// Licensed under the Eclipse Public License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License. You may
+// obtain a copy of the License at https://www.eclipse.org/legal/epl-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import core, { Class, Doc, Space, DocumentQuery, Ref, Storage, Tx, Account, TxCreateDoc, TxAddCollection, Emb, TxProcessor, makeEmb, Member, Hierarchy } from '@anticrm/core'
+
+export class SecurityStorage {
+  private readonly storage: Storage
+  private readonly hierarchy: Hierarchy
+  private readonly allowedSpaces: Map<Ref<Account>, Set<Ref<Space>>> = new Map<Ref<Account>, Set<Ref<Space>>>()
+  private readonly publicSpaces: Set<Ref<Space>> = new Set<Ref<Space>>()
+
+  constructor (storage: Storage, hierarchy: Hierarchy) {
+    this.storage = storage
+    this.hierarchy = hierarchy
+  }
+
+  async findAll<T extends Doc>(_class: Ref<Class<T>>, query: DocumentQuery<T>, userId: Ref<Account>): Promise<T[]> {
+    if (this.hierarchy.isDerived(_class, core.class.Space)) return await this.storage.findAll(_class, query)
+    const querySpace = (query as any).space as Ref<Space>
+    const spaces = new Set<Ref<Space>>([...this.publicSpaces, ...this.allowedSpaces.get(userId) ?? []])
+    if (querySpace !== undefined) {
+      if (spaces === undefined || spaces.size === 0) throw new Error('Access denied')
+      if (!spaces.has(querySpace)) throw new Error('Access denied')
+    } else {
+      if (spaces !== undefined) {
+        (query as any).space = [...spaces.values()]
+      }
+    }
+    return await this.storage.findAll(_class, query)
+  }
+
+  async tx (tx: Tx, userId: Ref<Account>): Promise<void> {
```

Comment:
Please use TxProcessor so you will not need to check for transactions, just delegate forward.